File: BlazorClient/Services/WeatherForecastCacheService.cs
using Rtm.BlazorClient.Models;

namespace Rtm.BlazorClient.Services;

public class WeatherForecastCacheService
{
    private readonly Lock _weatherLock = new();
    private WeatherForecastSeriesModel _weatherForecastSeriesModel =
        new() { WeatherForecasts = [], Timestamp = DateTime.Now };

    public event Action? OnChange;

    public WeatherForecastSeriesModel GetWeatherForecastSeries()
    {
        return _weatherForecastSeriesModel;
    }

    public void UpdateWeatherData(IEnumerable<WeatherForecastModel> newData)
    {
        lock (_weatherLock)
        {
            _weatherForecastSeriesModel = new WeatherForecastSeriesModel
            {
                WeatherForecasts = (newData ?? [])
                    .OrderBy(f => f.Date)
                    .ToList(),
                Timestamp = DateTime.Now,
            };
        }

        OnChange?.Invoke();
    }
}

File: NodeClient/src/services/weatherForecastCacheService.ts
import { WeatherForecastModel } from '../models/WeatherForecastModel';
import { WeatherForecastSeriesModel } from '../models/WeatherForecastSeriesModel';

let cachedWeatherForecasts: WeatherForecastSeriesModel;

export function setWeatherForecasts(data: WeatherForecastModel[]) {
  cachedWeatherForecasts = {
    weatherForecasts: data,
    timestamp: new Date(),
  };
}

export function getWeatherForecasts(): WeatherForecastSeriesModel {
  return cachedWeatherForecasts;
}
